{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rapidfuzz\n",
    "import Bio.SeqIO\n",
    "import tqdm.auto\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_to_references(sequences, reference_sequences):\n",
    "    \"\"\"Compare sequences to reference sequences.\n",
    "\n",
    "    Args:\n",
    "        sequences (dict): Dictionary of {seq_id: sequence}.\n",
    "        reference_sequences (dict): Dictionary of {seq_id: sequence}.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Dataframe with the following columns:\n",
    "            seqid (str): Sequence ID.\n",
    "            ref_seqid (str): Reference sequence ID.\n",
    "            edit_distance (int): Edit distance to reference sequence.\n",
    "            similarity (float): Normalized similarity to reference sequence.\n",
    "            length (int): Sequence length.\n",
    "    \"\"\"\n",
    "    \n",
    "    # get the reference sequence for each sequence, and calculate metrics\n",
    "    seq_ids = []\n",
    "    ref_seqids = []\n",
    "    edit_distances = []\n",
    "    similarities = []\n",
    "    lengths = []\n",
    "    for seq_id, seq in tqdm.auto.tqdm(sequences.items()):\n",
    "        _, _, ref_seqid = rapidfuzz.process.extractOne(seq, reference_sequences)\n",
    "        seq_ids.append(seq_id)\n",
    "        ref_seqids.append(ref_seqid)\n",
    "        lengths.append(len(seq))\n",
    "        edit_distances.append(rapidfuzz.distance.Levenshtein.distance(seq, reference_sequences[ref_seqid]))\n",
    "        similarities.append(rapidfuzz.distance.Levenshtein.normalized_similarity(seq, reference_sequences[ref_seqid]))\n",
    "\n",
    "    # create a dataframe with the results\n",
    "    df = pd.DataFrame({\n",
    "        \"seqid\": seq_ids, \n",
    "        \"ref_seqid\": ref_seqids, \n",
    "        \"edit_distance\": edit_distances, \n",
    "        \"similarity\": similarities, \n",
    "        \"length\": lengths\n",
    "    })\n",
    "    return df\n",
    "\n",
    "\n",
    "ref_seqs = {record.id: str(record.seq) for record in Bio.SeqIO.parse(\"./design_files.fasta\", \"fasta\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# unmapped reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dfs = []\n",
    "\n",
    "# for i in range(1, 6+1):\n",
    "#     sequences = {record.id: str(record.seq) for record in Bio.SeqIO.parse(gzip.open(f\"./PCR{i}/unmapped.fq.gz\", \"rt\"), \"fastq\")}\n",
    "#     df = compare_to_references(sequences, ref_seqs)\n",
    "#     df[\"exp\"] = f\"PCR{i}\"\n",
    "#     dfs.append(df)\n",
    "\n",
    "# df = pd.concat(dfs).reset_index(drop=True)\n",
    "# df.to_csv(\"unmapped_data.csv\", index=False)\n",
    "df = pd.read_csv(\"unmapped_data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.histogram(\n",
    "    df, \n",
    "    x=\"edit_distance\", \n",
    "    facet_col=\"exp\", \n",
    "    facet_col_wrap=2, \n",
    "    facet_row_spacing=0.1,\n",
    "    facet_col_spacing=0.05,\n",
    "    range_x=[0, 120],\n",
    "    range_y=[0, 175],\n",
    ")\n",
    "\n",
    "fig.update_traces(marker=dict(line=dict(width=0)), xbins=dict(start=0, end=150, size=2))\n",
    "\n",
    "fig.update_xaxes(title_text=\"Edit distance\", row=1)\n",
    "fig.update_yaxes(title_text=\"Count\", col=1)\n",
    "fig.for_each_annotation(lambda a: a.update(text=\"\"))\n",
    "\n",
    "for i in range(1, 7):\n",
    "    row = 3 - (i-1) // 2\n",
    "    col = (i-1) % 2 + 1\n",
    "    fig.add_annotation(\n",
    "        x=20, \n",
    "        y=150, \n",
    "        text=f\"<b>PCR {i}</b><br>n = {len(df[df['exp'] == f'PCR{i}'])}\", \n",
    "        showarrow=False, \n",
    "        font=dict(size=28/3, family=\"Inter\"),\n",
    "        row=row,\n",
    "        col=col,\n",
    "    )\n",
    "\n",
    "\n",
    "fig.update_layout(\n",
    "    template=\"simple_white\",\n",
    "    font_family=\"Inter\",\n",
    "    legend_font_size=28/3,\n",
    "    margin=dict(l=0, r=10, t=10, b=0),\n",
    "    width=320,\n",
    "    height=300,\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    minor_ticks=\"outside\", \n",
    "    title_font_family=\"Inter\", \n",
    "    title_font_size=28/3, \n",
    "    tickfont_size=28/3, \n",
    ")\n",
    "fig.update_xaxes(\n",
    "    minor_ticks=\"outside\", \n",
    "    title_font_family=\"Inter\", \n",
    "    title_font_size=28/3, \n",
    "    tickfont_size=28/3, \n",
    ")\n",
    "fig.for_each_annotation(lambda a: a.update(\n",
    "    font_size=28/3,\n",
    "    font_family=\"Inter\",\n",
    "))\n",
    "fig.show()\n",
    "fig.write_image(\"unmapped_distances.svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# unassigned reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unassigned = pd.read_csv(\"./unassigned/scafstats.txt\", sep=\"\\t\", dtype={'#name': str})\n",
    "unassigned.rename(columns={\"#name\": \"seq_id\"}, inplace=True)\n",
    "unassigned = unassigned[[\"seq_id\", \"assignedReads\"]].copy()\n",
    "\n",
    "params = pd.read_csv(\"params.csv\", dtype={'seq_id': str})\n",
    "counts = pd.read_csv(\"abundance_by_experiment.csv\", dtype={'seq_id': str})\n",
    "total_df = pd.merge(unassigned, params, on=\"seq_id\")\n",
    "total_df = pd.merge(total_df, counts, on=\"seq_id\")\n",
    "total_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_df[\"totalCount\"] = total_df[[f\"PCR{i}\" for i in range(1, 6+1)]].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.scatter(\n",
    "    total_df, \n",
    "    x=\"assignedReads\", \n",
    "    y=\"totalCount\", \n",
    "    color=\"eff\",\n",
    "    color_continuous_scale=px.colors.diverging.RdYlBu,\n",
    "    range_color=[0.95,1.0],\n",
    ")\n",
    "\n",
    "fig.update_traces(marker=dict(size=2, opacity=0.5, line=dict(width=0)))\n",
    "\n",
    "fig.update_layout(coloraxis_colorbar=dict(\n",
    "    title=\"PCR eff.\",\n",
    "    title_font_family=\"Inter\", \n",
    "    title_font_size=28/3, \n",
    "    tickfont_size=28/3,\n",
    "    thicknessmode=\"pixels\", \n",
    "    thickness=20,\n",
    "    lenmode=\"pixels\", \n",
    "    len=200,\n",
    "    yanchor=\"top\", y=0.8,\n",
    "))\n",
    "\n",
    "fig.add_annotation(\n",
    "    x=1.2, \n",
    "    y=0.95, \n",
    "    xref=\"paper\",\n",
    "    yref=\"paper\",\n",
    "    text=f\"n = {total_df.assignedReads.sum()}<br>σ = {total_df.assignedReads.corr(total_df.totalCount, method=\"spearman\"):.2f}\", \n",
    "    showarrow=False, \n",
    "    font=dict(size=28/3, family=\"Inter\"),\n",
    "    xanchor=\"center\",\n",
    "    yanchor=\"top\",\n",
    ")\n",
    "lineannotations = {\n",
    "    0.01: [80, 3900],\n",
    "    0.05: [240, 3900],\n",
    "    0.1: [445, 3900],\n",
    "    0.2: [600, 2700],\n",
    "}\n",
    "for fac in [0.01, 0.05, 0.1, 0.2]:\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=[0, 1e6*fac], \n",
    "        y=[0, 1e6],\n",
    "        marker_color=\"#bdbdbd\",\n",
    "        mode=\"lines\",\n",
    "    ))\n",
    "    fig.add_annotation(\n",
    "        x=lineannotations[fac][0], \n",
    "        y=lineannotations[fac][1], \n",
    "        text=f\"{fac:.0%}\", \n",
    "        showarrow=False, \n",
    "        font=dict(size=28/3, family=\"Inter\", color=\"#bdbdbd\"),\n",
    "    )\n",
    "    print(fac, total_df.loc[total_df.assignedReads < total_df.totalCount*fac].shape[0] / total_df.shape[0])\n",
    "\n",
    "rangeannotations = {\n",
    "    (0.00, 0.01): [5, 3000],\n",
    "    (0.01, 0.05): [90, 3000],\n",
    "    (0.05, 0.1): [210, 2900],\n",
    "    (0.1, 0.2): [380, 2600],\n",
    "    (0.2, 1.0): [550, 900],\n",
    "}\n",
    "for (start, end), pos in rangeannotations.items():\n",
    "    fig.add_annotation(\n",
    "        x=pos[0], \n",
    "        y=pos[1], \n",
    "        text=f\"{total_df.loc[(total_df.assignedReads < total_df.totalCount*end) & (total_df.assignedReads >= total_df.totalCount*start)].shape[0] / total_df.shape[0]:.1%}\", \n",
    "        showarrow=False, \n",
    "        font=dict(size=20/3, family=\"Inter\", color=\"#bdbdbd\"),\n",
    "    )\n",
    "\n",
    "\n",
    "fig.data = fig.data[::-1]\n",
    "fig.update_xaxes(\n",
    "    title_text=\"Mapped reads without index-based assignment\",\n",
    "    range=[-10, 1.01*total_df.assignedReads.max()],\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    title_text=\"Total reads across sequencing run\",\n",
    "    range=[-100, 1.01*total_df.totalCount.max()],\n",
    ")\n",
    "fig.update_layout(\n",
    "    template=\"simple_white\",\n",
    "    font_family=\"Inter\",\n",
    "    legend_font_size=28/3,\n",
    "    margin=dict(l=0, r=10, t=10, b=0),\n",
    "    width=320,\n",
    "    height=300,\n",
    "    showlegend=False,\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    minor_ticks=\"outside\", \n",
    "    title_font_family=\"Inter\", \n",
    "    title_font_size=28/3, \n",
    "    tickfont_size=28/3, \n",
    ")\n",
    "fig.update_xaxes(\n",
    "    minor_ticks=\"outside\", \n",
    "    title_font_family=\"Inter\", \n",
    "    title_font_size=28/3, \n",
    "    tickfont_size=28/3, \n",
    ")\n",
    "fig.show()\n",
    "fig.write_image(\"unassigned_abundance.svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import svgutils.transform as sg\n",
    "fig = sg.SVGFigure(\"680px\", \"315px\")\n",
    "\n",
    "panel = sg.fromfile(f'unmapped_distances.svg').getroot()\n",
    "panel.moveto(0, 10)\n",
    "fig.append(panel)\n",
    "\n",
    "panel = sg.fromfile(f'unassigned_abundance.svg').getroot()\n",
    "panel.moveto(680-320, 10)\n",
    "fig.append(panel)\n",
    "\n",
    "txt = sg.TextElement(2, 10, \"a\", size=28/3, weight=\"bold\", font=\"Inter\")\n",
    "fig.append([txt])\n",
    "txt = sg.TextElement(680-320+2, 10, \"b\", size=28/3, weight=\"bold\", font=\"Inter\")\n",
    "fig.append([txt])\n",
    "\n",
    "fig.save(f'unmapped_composed.svg')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "standard-3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
